packages/nodes-base/nodes/ActionNetwork/resources/person.ts

```diff
@@ -0,0 +1,393 @@
+			},
+			{
+				name: 'add_tags',
+				displayName: 'Tags',
+				type: 'string',
+				default: '',
+			},
+			{
+				name: 'identifiers',
+				displayName: 'Custom IDs',
+				type: 'string',
+				default: '',
+			},
+			{
+				name: 'custom_fields',
+				displayName: 'Custom Fields',
+				values: [
+					{
+						displayName: 'Key',
+						name: 'key',
+						type: 'string',
+						description: 'Custom field name as defined in Action Network',
+						default: '',
+					},
+					{
+						displayName: 'Value',
+						name: 'value',
+						type: 'string',
+						description: 'Custom field value',
+						default: '',
+					},
+				],
+			},
+		],
+	},
+])
+
+// ...constructPersonSignupHelperPayload({
+// 	family_name: this.getNodeParameter('family_name', i),
+// 	given_name: this.getNodeParameter('given_name', i),
+// 	postal_addresses: person_fields.postal_addresses,
+// 	email_addresses: person_fields.email_addresses,
+// 	phone_numbers: person_fields.phone_numbers,
+// 	add_tags: person_fields.add_tags,
+// 	custom_fields: person_fields.custom_fields
+// })
+/**
+ * {
+	family_name,
+	given_name,
+	email_addresses,
+	add_tags,
+	phone_numbers,
+	custom_fields,
+	postal_addresses
+}: any
+ */
+export function createPersonSignupHelperObject(node: IExecuteFunctions) {
+	// @ts-ignore
+	const {
+		email_addresses,
+		add_tags,
+		identifiers,
+		phone_numbers,
+		custom_fields,
+		postal_addresses
+	} = node.getNodeParameter('new_person_fields', 0, {}) as any
+
+	if (!email_addresses.length) {
+		throw new Error("Please add at least one email address")
+	}
+
+	const body: any = {
+		identifiers,
+		person: {
+			family_name: node.getNodeParameter('family_name', 0),
+			given_name: node.getNodeParameter('given_name', 0),
+			email_addresses
+		},
+		add_tags
+	}
+
+	// @ts-ignore
+	if (phone_numbers?.length > 0) {
+		body.person.phone_numbers = phone_numbers
+	}
+	if (Object.keys(custom_fields || {}).length > 0) {
+		body.person.custom_fields = (custom_fields as any)?.reduce(
+			// @ts-ignore
+			(obj, { key, value }) => ({ ...obj, [key]: value }),
+			{} as any
+		)
+	}
+	if (postal_addresses !== undefined && postal_addresses?.length > 0) {
+		body.postal_addresses = postal_addresses.map(
+			// @ts-ignore
+			({ street_address, latitude, longitude, accuracy, ...a }) => {
+				return {
+					...a,
+					address_lines: street_address ? [street_address] : undefined,
+					location: { latitude, longitude, accuracy }
+				}
+			}
+		)
+	}
+
+	return body
+}
+
+// [DOCS](https://actionnetwork.org/docs/v2/person)
+// Scenario: Retrieving a collection of attendance resources (GET)
+// Scenario: Retrieving an individual attendance resource (GET)
+// Scenario: Creating a new attendance (POST)
+// Scenario: Modifying an attendance (PUT)
+
+export const fields = [
+	{
+		displayName: 'Operation',
+		name: 'operation',
+		type: 'options',
+		default: 'GET',
+		description: 'Operation to perform',
+		options: [
+			{
+				name: 'Get',
+				value: 'GET',
+			},
```

Comment:
Thinking about maybe clarifying the 'operations' so that they are more specific. There are a few `get` requests, for example. The docs specify distinct scenarios with different URLs, methods, body, params -- I might use that instead?

- Scenario: Retrieving a collection of person resources (GET)
- Scenario: Retrieving an individual person resource (GET)
- Scenario: Creating a new person (POST)
- Scenario: Modifying a person (PUT)

This would make the conditional logic below even smoother. You wouldn't have to check if `person_id === ''` in the UI for example.

For the n8n team: I'm slowly figuring out what a 'good practice' node architecture looks like, but it would be great to compile this down into a "node design best practices" guide. Things in my mind include:

- Collocate UI and logic to make it easy to maintain and edit
- Using the paradigm of "resources" and "operations", you should be able to create exclusive conditionals in which to show specific fields and do specific logic, like a particular API field. A good example is  `{ resource: People, operation: "List People" }` vs `{ resource: People, operation: GET, person_id: not required }`
- Don't be afraid to duplicate logic and UI if it makes things clearer
- Prefer that instead of trying to entangle different things in ways that makes the UI and parameters unpredictable

Reply:
[This wiki](https://github.com/MLH-Fellowship/n8n/wiki/Best-practices) might be useful.